test/unit/TestLanguagePipeline.hs

```diff
@@ -333,7 +333,34 @@ testLanguagePipeline =
             "checking a lambda with the wrong argument type"
             ( process
                 "(\\x:int. x + 2) : text -> int"
-                "1:1: Lambda argument has type annotation int, but expected argument type text"
+                "1:1: Lambda argument has type annotation `int`, but expected argument type `text`"
+            )
+        ]
+    , testGroup
+        "typechecking errors"
+        [ testCase
+            "applying a pair"
+            ( process
+                "(1,2) \"hi\""
+                "1:1: Type mismatch:\n  From context, expected `(1, 2)` to be a function,\n  but it is actually a pair"
+            )
+        , testCase
+            "providing a pair as an argument"
+            ( process
+                "(\\x:int. x + 1) (1,2)"
+                "1:17: Type mismatch:\n  From context, expected `(1, 2)` to have type `int`,\n  but it is actually a pair"
+            )
+        , testCase
+            "mismatched if branches"
+            ( process
+                "if true {grab} {}"
+                "1:16: Type mismatch:\n  From context, expected `noop` to have type `cmd text`,\n  but it actually has type `cmd unit`"
+            )
+        , testCase
+            "definition with wrong result"
+            ( process
+                "def m : int -> int -> int = \\x. \\y. {3} end"
+                "1:37: Type mismatch:\n  From context, expected `{3}` to have type `int`,\n  but it is actually a delayed expression"
```

Comment:
Could you please  add a test case for two functions?

It seems to be handled based on the `cmd` test case above, but it would be nice to be sure we don't say:
> expected `f` to be a _function_, but it is actually a _function!_

Reply:
I added a couple test cases involving incompatible functions to show what happens.  We will never say "expected f to be a function, but it is actually a function", for two reasons: first, if we are trying to unify two types and they are both function types, then we will proceed to unifying their input and output types, so we would never fail at the point where we have two function types.  Second, we only say "blah is a function" when it specifically has a type of the form `u1 -> u2` where `u1` and `u2` are unification variables.  Any two such types will be guaranteed to be unifiable.